trieste/acquisition/function.py

```diff
@@ -676,3 +676,243 @@ def batch_ei(at: TensorType) -> TensorType:
+        :return: An acquisition function builder that selects the model and dataset specified by
+            ``tag``, as defined in :meth:`prepare_acquisition_function`.
+        """
+        single_builder = self
+
+        class _Anon(GreedyAcquisitionFunctionBuilder):
+            def prepare_acquisition_function(
+                self,
+                datasets: Mapping[str, Dataset],
+                models: Mapping[str, ProbabilisticModel],
+                pending_points: Optional[TensorType] = None,
+            ) -> AcquisitionFunction:
+                return single_builder.prepare_acquisition_function(
+                    datasets[tag], models[tag], pending_points=pending_points
+                )
+
+            def __repr__(self) -> str:
+                return f"{single_builder!r} using tag {tag!r}"
+
+        return _Anon()
+
+    @abstractmethod
+    def prepare_acquisition_function(
+        self,
+        dataset: Dataset,
+        model: ProbabilisticModel,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data to use to build the acquisition function.
+        :param model: The model over the specified ``dataset``.
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+
+
+class LocallyPenalizedExpectedImprovement(SingleModelGreedyAcquisitionBuilder):
+    r"""
+    Builder of the acquisition function maker for greedily collecting batches by local
+    penalization.  The resulting :const:`AcquisitionFunctionMaker` takes in a set of pending
+    points and returns the expected improvment acquisiton function penalized around those points.
+
+    Local penalization allows us to perform batch Bayesian optimization with a standard (non-batch)
+    acqusition function. By iteratively building a batch of points by maximizing an acquisition
+    function that is clipped around the values correponding to locations close to the already
+    chosen (pending) points, local penalization provides diverse batches of canidadate points.
+
+    Local penalization is applied to the acquisition function multiplicatively. However, to
+    improve numerical stability, we perfom additive penalization in a log space.
+
+    For our penalization function, we implement the soft penalization strategy
+    of :cite:`Gonzalez:2016`, where an estimate of the objective function's Lipschitz
+    constant is used to control the size of penalization.
+
+    The Lipschitz constant and additional penalization parameters are estimated once
+    when first preparing the acquisition function with no pending points. These estaimtes
+    are reused for all subsequent function calls.
+    """
+
+    def __init__(self, search_space: SearchSpace, grid_size: int = 500):
```

Comment:
grid_size is misleading: we're not using a grid (search_space.sample() returns uniformly distributed samples). Plus, grid_size could be understood as `sample_size = grid_size ^ dim`